lib/FetchEnvs.js

```diff
@@ -228,42 +227,71 @@ module.exports = class FetchEnvs {
     const matchLabelsQS = labelSelectors(matchLabels);
 
     const cacheKey = [apiVersion, kind, namespace, name].join('/');
-    // Note: Using `matchLabels` will always result in a kube api call, label-based queries cannot use the resourceCache
-    if( !matchLabelsQS && this.resourceCache.has(cacheKey) ) {
-      response = this.resourceCache.get(cacheKey);
-    }
-    else {
-      const krm = await this.kubeClass.getKubeResourceMeta(apiVersion, kind, 'update');
 
-      if (krm) {
-        try {
+    let kubeError = ERR_NODATA;
+    let response;
+
+    if( typeof matchLabelsQS === OBJECT ) {
+      // MatchLabels queries are not cached (though the resulting resources are cached), always make the kube api call
+      try {
+        const krm = await this.kubeClass.getKubeResourceMeta(apiVersion, kind, 'update');
+
+        if (krm) {
           response = await this.api({
             uri: krm.uri({ namespace, name }),
             json: true,
             qs: matchLabelsQS
           });
-          // Note: cache here only if getting a single resource
-          if( response?.data && !response?.items ) {
-            this.resourceCache.set(cacheKey, response);
+          // Cache multiple resources
+          if( response?.items ) {
+            response.items.forEach(function (item) {
+              const cacheKey = [item.apiVersion, item.kind, item.metadata.namespace, item.metadata.name].join('/');
+              this.resourceCache.set(cacheKey, item);
+            }, this);
           }
-        } catch (error) {
-          kubeError = error;
         }
+      } catch (error) {
+        kubeError = error;
+      }
+    }
+    // Single-resource queries are cacheable.  If in the cache, use it.
+    else if( this.resourceCache.has( cacheKey ) ) {
+      response = this.resourceCache.get( cacheKey );
+    }
+    // Single-resource queries are cacheable.  If not in the cache, start an api call to populate the cache if needed, wait for it to finish, then use it from the cache.
+    else {
+      if( !singleResourceQueryCache[cacheKey] ) {
```

Comment:
Could we call #genericMapRef here to get the map, avoiding what seems like duplicate code?

Reply:
`genericKeyRef can't directly call `genericMapRef` without additional refactoring, but combined the common function into `getSingleResource` function used in both places.